test/powershell/engine/Api/MultithreadedEngine.Tests.ps1

```diff
@@ -0,0 +1,198 @@
+        }
+
+        It 'cannot invoke a single script asynchronously in a runspace that has not been opened' {
+            $rs = [runspacefactory]::CreateRunspace()
+            $ps = [powershell]::Create($rs)
+            $r = $ps.AddScript('@(1..10).foreach{Start-Sleep -Milliseconds 500}').InvokeAsync()
+            # This test is designed to fail. You cannot invoke PowerShell asynchronously
+            # in a runspace that has not been opened.
+            $r.IsFaulted | Should -Be $true
+            $r.Exception.InnerException -is [System.Management.Automation.Runspaces.InvalidRunspaceStateException] | Should -Be $true
+            $r.Exception.InnerException.CurrentState | Should -Be 'BeforeOpen'
+            $r.Exception.InnerException.ExpectedState | Should -Be 'Opened'
+        }
+
+        It 'cannot invoke a single script asynchronously in a runspace that is busy' {
+            $rs = [runspacefactory]::CreateRunspace()
+            $rs.Open()
+            $psBusy = [powershell]::Create($rs)
+            $r = $psBusy.AddScript('@(1..5).foreach{Start-Sleep -Milliseconds 500}').InvokeAsync()
+            $rs.RunspaceAvailability | Should -Be 'Busy'
+            $ps = [powershell]::Create($rs)
+            $sb = {
+                InvokeAsyncHelper -PowerShell $ps -Wait
+            }
+            # This test is designed to fail. You cannot invoke PowerShell asynchronously
+            # in a runspace that is busy, because pipelines cannot be run concurrently.
+            $err = { $sb.Invoke() } | Should -Throw -ErrorId 'AggregateException' -PassThru
+            GetInnerErrorId -Exception $err.Exception | Should -Be 'InvalidOperation'
+            $count = 0
+            while (-not $r.IsCompleted -and $count -lt 10) {
+                Start-Sleep -Milliseconds 500
+                $count++
+            }
+        }
+
+        It 'cannot invoke a single script asynchronously in the current runspace' {
+            $ps = [powershell]::Create('CurrentRunspace')
+            $sb = {
+                InvokeAsyncHelper -PowerShell $ps -Wait
+            }
+            # This test is designed to fail. You cannot invoke PowerShell asynchronously
+            # in the current runspace because nested PowerShell instances cannot be
+            # invoked asynchronously
+            $err = { $sb.Invoke() } | Should -Throw -ErrorId 'AggregateException' -PassThru
+            GetInnerErrorId -Exception $err.Exception | Should -Be 'InvalidOperation'
+        }
+    }
+
+    Context 'PowerShell::InvokeAsync - Multiple script tests' {
+        It 'can invoke multiple scripts asynchronously' {
+            $sb1 = [scriptblock]::Create("@(1,3,5,7,9,11,13,15,17,19)${sbStub}")
+            $sb2 = [scriptblock]::Create("@(2,4,6,8,10,12,14,16,18,20)${sbStub}")
+            $r1 = [powershell]::Create().AddScript($sb1).InvokeAsync()
+            $r2 = [powershell]::Create().AddScript($sb2).InvokeAsync()
```

Comment:
It's better to pass string directly to `AddScript`.

Reply:
Done.